peers/topicpool_test.go

```diff
@@ -328,3 +327,99 @@ func (s *TopicPoolSuite) TestNewTopicPoolInterface() {
 	cacheTP := newCacheOnlyTopicPool(tp, &testTrueVerifier{})
 	s.IsType(&cacheOnlyTopicPool{}, cacheTP)
 }
+
+func (s *TopicPoolSuite) TestIgnoreInboundConnection() {
+	s.topicPool.limits = params.NewLimits(0, 0)
+	s.topicPool.maxCachedPeers = 0
+	peer1 := discv5.NewNode(discv5.NodeID{1}, s.peer.Self().IP, 32311, 32311)
+	s.topicPool.processFoundNode(s.peer, peer1)
+	s.Contains(s.topicPool.pendingPeers, peer1.ID)
+	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer1.ID))
+	s.Contains(s.topicPool.pendingPeers, peer1.ID)
+	s.False(s.topicPool.pendingPeers[peer1.ID].dismissed)
+	s.NotContains(s.topicPool.connectedPeers, peer1.ID)
+}
+
+func (s *TopicPoolSuite) TestConnectedButRemoved() {
+	s.topicPool.limits = params.NewLimits(0, 0)
+	s.topicPool.maxCachedPeers = 1
+	peer1 := discv5.NewNode(discv5.NodeID{1}, s.peer.Self().IP, 32311, 32311)
+	s.topicPool.processFoundNode(s.peer, peer1)
+	s.Contains(s.topicPool.pendingPeers, peer1.ID)
+	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer1.ID))
+	s.Contains(s.topicPool.connectedPeers, peer1.ID)
+	s.False(s.topicPool.ConfirmDropped(s.peer, discover.NodeID(peer1.ID)))
+	s.False(s.topicPool.pendingPeers[peer1.ID].added)
+}
+
+func TestServerIgnoresInboundPeer(t *testing.T) {
+	topic := discv5.Topic("cap=cap1")
+	limits := params.NewLimits(0, 0)
+	cache, err := newInMemoryCache()
+	require.NoError(t, err)
+	topicPool := newTopicPool(nil, topic, limits, 100*time.Millisecond, 200*time.Millisecond, cache)
+	topicPool.running = 1
+	topicPool.maxCachedPeers = 0
+
+	whisper := whisperv6.New(nil)
+	srvkey, err := crypto.GenerateKey()
+	require.NoError(t, err)
+	server := &p2p.Server{
+		Config: p2p.Config{
+			MaxPeers:    1,
+			Name:        "server",
+			ListenAddr:  ":0",
+			PrivateKey:  srvkey,
+			NoDiscovery: true,
+			Protocols:   whisper.Protocols(),
+		},
+	}
+	require.NoError(t, server.Start())
+	clientkey, err := crypto.GenerateKey()
+	require.NoError(t, err)
+	client := &p2p.Server{
+		Config: p2p.Config{
+			MaxPeers:    1,
+			Name:        "client",
+			ListenAddr:  ":0",
+			PrivateKey:  clientkey,
+			NoDiscovery: true,
+			Protocols:   whisper.Protocols(),
+		},
+	}
+	require.NoError(t, client.Start())
+
+	// add peer to topic pool, as if it was discovered.
+	// it will be ignored due to the limit and added to a table of pending peers.
+	clientID := discv5.NodeID(client.Self().ID)
+	topicPool.processFoundNode(server, discv5.NewNode(
+		clientID,
+		client.Self().IP,
+		client.Self().UDP, client.Self().TCP))
+	require.Contains(t, topicPool.pendingPeers, clientID)
+	require.False(t, topicPool.pendingPeers[clientID].added)
+
+	// connect to a server from client. client will be an inbound connection for a server.
+	client.AddPeer(server.Self())
+	errch := helpers.WaitForPeerAsync(server, client.Self().String(), p2p.PeerEventTypeAdd, 5*time.Second)
```

Comment:
This can be put before `client.AddPeer(server.Self())` to make sure we won't miss the event.